src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -499,352 +527,305 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
                 // Builds the compilation matching the project state. In the process of building, also
                 // produce in progress snapshots that can be accessed from other threads.
                 // </summary>
-                async Task<FinalState> BuildFinalStateAsync(
-                    SolutionCompilationState compilationState,
-                    CancellationToken cancellationToken)
+                async Task<FinalCompilationTrackerState> BuildFinalStateAsync()
                 {
                     cancellationToken.ThrowIfCancellationRequested();
 
                     var state = ReadState();
 
-                    return state switch
-                    {
-                        // if we already have a compilation, we must be already done!  This can happen if two
-                        // threads were waiting to build, and we came in after the other succeeded.
-                        FinalState finalState
-                            => finalState,
+                    // if we already have a compilation, we must be already done!  This can happen if two
+                    // threads were waiting to build, and we came in after the other succeeded.
+                    if (state is FinalCompilationTrackerState finalState)
+                        return finalState;
```

Comment:
This comment, although not new, seems terrifying -- don't we always have a lock around this now? Is this now moot since you deleted the takeLock thing in an earlier PR?

Reply:
No, this is totally normal.  Note: i got rid of the takeLock flag because we *always* take the lock through all codepaths.

But the logic is:

1. try to check the state outside of hte lock.  if a final-state we're done.
2. take lock.
3. check the state again.  if a final state we're done
4. compute the final state.

this is appropraite :)